source/common/http/codes.h

```diff
@@ -8,62 +8,120 @@
+  Stats::StatName from_zone_;
+  Stats::StatName to_zone_;
+  bool upstream_canary_;
+};
+
+struct CodeStats::ResponseTimingInfo {
+  Stats::Scope& global_scope_;
+  Stats::Scope& cluster_scope_;
+  Stats::StatName prefix_;
+  std::chrono::milliseconds response_time_;
+  bool upstream_canary_;
+  bool internal_request_;
+  Stats::StatName request_vhost_name_;
+  Stats::StatName request_vcluster_name_;
+  Stats::StatName from_zone_;
+  Stats::StatName to_zone_;
+};
+
 class CodeStatsImpl : public CodeStats {
 public:
-  ~CodeStatsImpl() override = default;
+  explicit CodeStatsImpl(Stats::SymbolTable& symbol_table);
 
   // CodeStats
-  void chargeBasicResponseStat(Stats::Scope& scope, const std::string& prefix,
+  void chargeBasicResponseStat(Stats::Scope& scope, Stats::StatName prefix,
                                Code response_code) const override;
   void chargeResponseStat(const ResponseStatInfo& info) const override;
   void chargeResponseTiming(const ResponseTimingInfo& info) const override;
 
 private:
   friend class CodeStatsTest;
 
-  /**
-   * Strips any trailing "." from a prefix. This is handy as most prefixes
-   * are specified as a literal like "http.", or an empty-string "". We
-   * are going to be passing these, as well as other tokens, to join() below,
-   * which will add "." between each token.
-   *
-   * TODO(jmarantz): remove all the trailing dots in all stat prefix string
-   * literals. Then this function can be removed.
-   */
+  void incCounter(Stats::Scope& scope, const std::vector<Stats::StatName>& names) const;
+  void incCounter(Stats::Scope& scope, Stats::StatName a, Stats::StatName b) const;
+  void recordHistogram(Stats::Scope& scope, const std::vector<Stats::StatName>& names,
+                       uint64_t count) const;
+
+  class RequestCodeGroup {
+  public:
+    RequestCodeGroup(absl::string_view prefix, CodeStatsImpl& code_stats);
+
+    Stats::StatName statName(Code response_code);
+
+  private:
+    // Use an array of atomic pointers to hold StatNameStorage objects for
+    // every conceivable HTTP response code. In the hot-path we'll reference
+    // these with a null-check, and if we need to allocate a symbol for a
+    // new code, we'll take a mutex to avoid duplicate allocations and
+    // subsequent leaks. This is similar in principle to a ReaderMutexLock,
+    // but should be faster, as ReaderMutexLocks appear to be too expensive for
+    // fine-grained controls. Another option would be to use a lock per
+    // stat-name, which might have similar performance to atomics with default
+    // barrier policy.
+    //
+    // We don't allocate these all up front during construction because
+    // SymbolTable greedily encodes the first 128 names it discovers in one
+    // byte. We don't want those high-value single-byte codes to go to fully
+    // enumerating the 4 prefixes combined with HTTP codes that are seldom used,
+    // so we allocate these on demand.
```

Comment:
The codes object is global to the server, right? Can we talk about that a bit in the comment? And also how many of these groups there are?

Reply:
Added a note that the Codes object is global to the server. I folded the RequestCodeGroup into CodeStats since there's only one.